Expose left/right panning as a pref

keianhzo opened the pull request:
Closes #4639

This PR exposes Left/Right panning as an accessibility setting. Thing to note:
- We no longer use the `audioOutputMode` preference as it has been outlined in the code comments as confusing, now we have a `disableLeftRightPanning` setting and the `audiomode` command updates that pref.
https://github.com/mozilla/hubs/blob/fcf78a30652cb745f90d8e20159ade3031157d2a/src/message-dispatch.js#L179-L181
- This PR removes the initial panning mode pref resetting that was happening before.
https://github.com/mozilla/hubs/blob/fcf78a30652cb745f90d8e20159ade3031157d2a/src/systems/audio-settings-system.js#L8-L17
- Also we no longer log in the console when the mode is changed (we don't do that for any other pref either so it kind of makes sense now that it has graduated to pref)

johnshaughnessy commented on src/systems/sound-effects-system.js:
```diff
@@ -138,7 +138,7 @@ export class SoundEffectsSystem {
     if (!audioBuffer) return null;
 
-    const disablePositionalAudio = isSafari() || window.APP.store.state.preferences.audioOutputMode === "audio";
+    const disablePositionalAudio = isSafari() || window.APP.store.state.preferences.disableLeftRightPanning === true;
```
Can `window.APP.store.state.preferences.disableLeftRightPanning === true` be reduced to `window.APP.store.state.preferences.disableLeftRightPanning` here?

keianhzo replied:
yep, I think so, we are only using it for the ternary. Updated.

johnshaughnessy commented on src/storage/store.js:
```diff
@@ -96,7 +96,6 @@ export const SCHEMA = {
         preferredMic: { type: "string" },
         preferredCamera: { type: "string" },
         muteMicOnEntry: { type: "bool" },
-        audioOutputMode: { type: "string" },
```
If you remove this key from the store, but it was set in a user's local storage because in their most recent setting they set their `audioOutputMode` pref (via the command), then when the user loads the page, the data in their local storage will cause them to fail the schema validation and the page will fail to load.

Before schema validation runs the first time, we need to delete this entry from users' local storage. Alternatively, we can leave it in the schema definition for a while (with code anywhere that deletes the stored value) with a comment that says, "Deprecated, to be removed once we're confident local storage has been cleaned up."

I think I like the first option better, where before doing schema validation we delete now-deprecated keys if they exist.

keianhzo replied:
I think we switched the preferences store to accept additional properties here:
https://github.com/mozilla/hubs/pull/4275/commits/e865eb78812980297b6ce6fae5236e3b737b0911#r636548037
That's why it doesn't really crash (and also why it doesn't crash even though I forgot to add the actual new `disableLeftRightPanning` pref 😬) I wonder if we should write cleanup code though so old prefs are removed and the store doesn't hold stall prefs.

johnshaughnessy replied:
Oh, great!

johnshaughnessy commented on src/message-dispatch.js:
```diff
@@ -174,20 +174,9 @@ export default class MessageDispatch extends EventTarget {
-            // and even if it's not, "audio" is a weird name for the "audioOutputMode" that means
-            // "stereo" / "not panner".
-            preferences: { audioOutputMode: shouldEnablePositionalAudio ? "panner" : "audio" }
+            preferences: { disableLeftRightPanning: args[0] === "panner" ? false : true }
```
We can remove this chat command

keianhzo replied:
I guess there is no point on keeping that around for compat. Removed.

johnshaughnessy commented on src/components/avatar-audio-source.js:
```diff
@@ -100,11 +100,11 @@ AFRAME.registerComponent("avatar-audio-source", {
     this.createAudio();
 
-    let audioOutputModePref = APP.store.state.preferences.audioOutputMode;
+    let disableLeftRightPanningPref = APP.store.state.preferences.disableLeftRightPanning === true;
```
I think you can drop `=== true`.

keianhzo replied:
In this case (same as in media-video) I think we want to only recreate in case it went from undefined/false <-> true and not if it changed from undefined <-> false

keianhzo replied:
Thinking about it again now that we don't have the chat command, I don't think there is a case where this goes from `undefined` to `false` so I think is safe to remove that comparison.

johnshaughnessy reviewed (changes requested):
My only major concern has to do with removing keys from the `store`. If a user has a key-value pair for `audioOutputMode`, this change will cause the schema validation to fail, which is an error we currently do not catch or recover from.

keianhzo commented:
@johnshaughnessy I've fixed all the issues and added a final commit to always remove unsupported properties from the store after validating. As we don't want to accept unsupported properties I think this should help keep the store clean.

johnshaughnessy commented on src/storage/store.js:
```diff
@@ -360,12 +359,13 @@ export default class Store extends EventTarget {
-      throw new Error(`Write to store failed schema validation.`);
+      errors.forEach(error => {
+        delete error.instance[error.argument];
+      });
```
It seems like we should at least log these errors. Also, is this change necessary/desirable given the `allowAdditionalProps` change you referred to in an earlier comment.

keianhzo replied:
First comment here https://github.com/mozilla/hubs/pull/4275/files#r636548037 talked about cleaning up the store to avoid stall prefs to stick around forever. I thought it was a good time for adding that.

johnshaughnessy commented on src/update-audio-settings.js:
```diff
@@ -36,7 +36,7 @@ export function getCurrentAudioSettings(el) {
   const zoneSettings = APP.zoneOverrides.get(el);
   const preferencesOverrides =
-    APP.store.state.preferences.audioOutputMode === "audio" ? { audioType: AudioType.Stereo } : {};
+    APP.store.state.preferences.disableLeftRightPanning === true ? { audioType: AudioType.Stereo } : {};
```
We can lose `=== true`
